import grpc
import logging
from google.protobuf.json_format import ParseDict, MessageToDict

# Generated code imports
# (ê²½ë¡œ ë¬¸ì œ í•´ê²°ì„ ìœ„í•´ try-except í˜¹ì€ ìƒëŒ€ ê²½ë¡œ ì¡°ì • í•„ìš”. ì—¬ê¸°ì„  ì ˆëŒ€ ê²½ë¡œ ê°€ì •)
try:
    from core.plugin.generated import gopedia_pb2, gopedia_pb2_grpc
except ImportError:  # pragma: no cover - only hit in local dev/tests without proto stubs
    gopedia_pb2 = None  # type: ignore[assignment]
    gopedia_pb2_grpc = None  # type: ignore[assignment]
from core.plugin.interface import PluginClientInterface
from domain.schemas.plugin import PluginPayload, PluginResult

logger = logging.getLogger(__name__)

class GrpcPluginClient(PluginClientInterface):
    """
    ì‹¤ì œ gRPC í†µì‹ ì„ ë‹´ë‹¹í•˜ëŠ” êµ¬í˜„ì²´.
    Pydantic Model <-> Protobuf Struct ë³€í™˜ì„ ìˆ˜í–‰í•©ë‹ˆë‹¤.
    """

    def __init__(self):
        self.channel = None
        self.stub = None

    async def connect(self, service_address: str) -> None:
        """ë¹„ë™ê¸° ì±„ë„ ìƒì„±"""
        if gopedia_pb2_grpc is None:
            raise RuntimeError(
                "gRPC stubs are missing. Run `python -m grpc_tools.protoc ...` "
                "to generate core.plugin.generated modules before connecting."
            )
        self.channel = grpc.aio.insecure_channel(service_address)
        self.stub = gopedia_pb2_grpc.PluginServiceStub(self.channel)
        logger.info(f"ğŸ”Œ Connected to Plugin Service at {service_address}")

    async def execute(self, payload: PluginPayload) -> PluginResult:
        if not self.stub:
            raise ConnectionError("Plugin client is not connected.")
        if gopedia_pb2 is None:
            raise RuntimeError(
                "gRPC stubs are missing. Run the protoc generation step before executing."
            )

        # 1. Pydantic -> Protobuf Request ë³€í™˜
        req = gopedia_pb2.PluginRequest(
            context_urn=payload.context_urn or "",
            operation=payload.operation
        )
        
        # JSON(Dict)ì„ Protobuf Structë¡œ ë³€í™˜
        if payload.params:
            ParseDict(payload.params, req.params)
        if payload.data_snapshot:
            ParseDict(payload.data_snapshot, req.payload)

        try:
            # 2. gRPC Call (Async)
            response = await self.stub.Execute(req)

            # 3. Protobuf Response -> Pydantic ë³€í™˜
            result_data = MessageToDict(response.data) if response.data else None
            
            return PluginResult(
                success=response.success,
                error_message=response.error_message,
                data=result_data
            )

        except grpc.RpcError as e:
            logger.error(f"gRPC Error: {e.code()} - {e.details()}")
            return PluginResult(success=False, error_message=f"RPC Error: {e.details()}")

    async def close(self):
        if self.channel:
            await self.channel.close()